### DISCUSSION TITLE
MIR version 1.0.0

### DISCUSSION THREAD

**Discussion by vnmakarov** [vnmakarov/mir#398] [General]
I am glad to announce the release of MIR version 1.0.0.

This release is based on the `bbv` branch, which has been in use by
some people for some time.  Branch `bbv` became `master`.  The
previous branch `master` became `v0_master`.

Here are the major changes compared to the previous
version, 0.2.0:

* Major **rewriting of the optimization pipeline** of the MIR-generator:
  * This results in better-generated code, which yields a **14%** average performance improvement and
    a **16%** geometric mean improvement on a set of benchmarks found in the `c-benchmarks` directory.
* Introduction of experimental support for **lazy basic block versioning**, which can be utilized
  to implement a better JIT for dynamic programming languages:
  * Machine code for basic blocks can now be generated lazily upon their first execution.
  * Lazy basic block generation assumes lazy generation of functions.
* New features on the MIR and C level:
  * MIR instructions to take and use address of MIR variables to simplify code generation of MIR code
  * MIR and C variables bound to machine registers.
  * Implementation of a new C builtin `__builtin_expect`
  * MIR instructions and corresponding builtins in C related to integer overflow.
  * Indirect jumps in MIR and labels as values in C.
  * MIR instructions and corresponding builtins in C to implement call ABI based on jumps.
  * Properties for basic block versioning.
* Removed features:
  * The PPC64BE target has been removed as the last big-endian CPU (power7) became too old.
    MIR now supports only PPC64LE (little-endian).
  * Parallel generation of functions has been removed as it does not integrate well with lazy function
    and bbv code generation. Lazy function generation and lazy basic block generation work
    at practically the same speed.
    * Users can implement parallel generation using different contexts if desired.
    * This also means that the MIR-generator API has changed slightly,
      hence the change in the MIR API version was upgraded from 0 to 1.
    * The `c2m` driver still supports parallel parsing of different C files.

<hr /><em>This discussion was created from the release <a href='https://github.com/vnmakarov/mir/releases/tag/v1.0.0'>MIR version 1.0.0</a>.</em>

**Comment by dibyendumajumdar** [comment]
Just flagging that I get an assertion failure in Ravi when using the latest. Not investigated yet.

**Comment by iacore** [comment]
@vnmakarov Do you plan to make the code more readable by explaining what it does with comments and write ups? Then, more people can chime in on the effort.

Zig allows very modular code, but the language is still evolving. If I were to implement further features, I would implement new code in Zig, then detangle (the reverse of complecting) C code into smaller modules.

Zig's dependent type system allows you to write very succinct code that generates other code that depends on the type of data being operated on. It is like reflection + hygienic macro.

Please consider using Zig so the code is better organized and more readable. https://ziglang.org/documentation/0.15.2/

**Comment by rempas** [comment]
Wow! I've run the latest commit and in the benchmarks, I got an average of 0.81x of "GCC -03" while "Clang -03" got 0.91x! This is damn impressive! And the big losses were Arrays (especially) and some others. In some tests, Mir was even faster than GCC!

Do you plan to ever come back?

**Reply by cyrilmhansen** [reply]
I feel I should submit your comment to hacker news to see if it sticks.
Last real MIR discussion on HN was two years ago. A new use case would be better than just an updated benchmark.
Is there a public list of projects based on MIR somewhere ? Codex "wrote" a retro BASIC interpreter with me based on MIR and got stuck when I tried to get pluggable floating point implementations. Basically I ran out of crazy ideas to implement and was not motivated enough to polish anything.

I suspect with the right documentation and tooling, Gemini 3.0 pro could "implement" "almost any" language based on MIR in "a few hours". A video could be very effective marketing.

**Reply by rempas** [reply]
I don't use HN but, if you think that it will help MIR get traction, that's cool! I am not aware of a list with projects that use MIR but, want to use it for the programming language I make, for the debug mode. I want to do hot module replacement and MIR seems to be able to do that, in my understanding! And I do plan to make it a big thing and not just a toy language for fun. So, MIR will get a good promotion out of it.

I also saw your fork of MIR, and I was planning to contact you in the future but since you added this reply, could I ask what is changed? If you fixed bugs and added features, I'd love to use your fork and start reading and understand the code from there. I won't start right yet, in fact, It might take a couple of months because my life is not in the best state right now, and I must focus on other things, but, we could exchange some ideas, no problem!
